40 add pagination projects page

mattwr18 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

#### Issue addressed
<!-- include `fixes #<issue-number>` -->
<!-- the relevant issue in https://github.com/AgileVentures/WebsiteOne-FE/issues  -->

fixes #40 

#### Screenshots (if appropriate):
<!-- please include screenshots of any changes to the UI for quick review  -->
<!-- please show how things look on both desktop and mobile  -->

#### Testing
<!-- Remember you must see any new tests you created (or old ones you changed) -->
<!-- fail as well as pass in order to ensure they are working -->
<!-- Unsure how to test? - please see https://github.com/AgileVentures/WebsiteOne-FE/blob/develop/CONTRIBUTING.md#git-and-github-->

mattwr18 committed:
Merge pull request #6 from AgileVentures/code_of_conduct

Add multiple files missing on creation

mattwr18 committed:
Add UsersList container, test (#7)

* Add UsersList container, test

fixes #5

dkuku committed:
semantic-ui assets (#13)

* semantic-ui assets

* exclude assets folders from coverage report

* add package.json

* Remove separate components from repo

* remove unused themes

dkuku committed:
Install semantic ui (#14)

* add path module

* add build cmd

* add new line

* fix style loader

* config files and manual for customizing semantic ui

* keep only changes in site.variables

* removed unused import

* removed unused import from tests

* add semantic folder to gitignore

* add deleted test

* exclude assets folders from coverage report

* added assets

* clean assets

* Remove package-lock

* Readme modification

dkuku committed:
Homepage (#15)

* add path module

* add build cmd

* add new line

* fix style loader

* config files and manual for customizing semantic ui

* keep only changes in site.variables

* removed unused import

* removed unused import from tests

* add semantic folder to gitignore

* add deleted test

* exclude assets folders from coverage report

* added assets

* clean assets

* add source files

* add images files

* add fixtures files

* fix jest failing on svg import

* rename page title

* render hemepage

* cleaned tests

* remove unused import from index page

* Remove extra space

mattwr18 committed:
9 add user and pagination components (#16)

* Add multiple files missing on creation

- Add Issue and Pull Request template
- Add License
- Add Code Of conduct

* Add UsersList container, test

fixes #5

* Bg fix style loader (#8)

* add path module

* add build cmd

* add new line

* fix style loader

* Add new lines at end of files

* Add pagination, user component

* Remove logic for extra links, import connected component

- We don't have 42 pages right now, as we're only getting 100 users
- copy and paste mistake

* Add Grid, styling

* semantic-ui assets (#13)

* semantic-ui assets

* exclude assets folders from coverage report

* add package.json

* Remove separate components from repo

* remove unused themes

* Install semantic ui (#14)

* add path module

* add build cmd

* add new line

* fix style loader

* config files and manual for customizing semantic ui

* keep only changes in site.variables

* removed unused import

* removed unused import from tests

* add semantic folder to gitignore

* add deleted test

* exclude assets folders from coverage report

* added assets

* clean assets

* Remove package-lock

* Readme modification

* Homepage (#15)

* add path module

* add build cmd

* add new line

* fix style loader

* config files and manual for customizing semantic ui

* keep only changes in site.variables

* removed unused import

* removed unused import from tests

* add semantic folder to gitignore

* add deleted test

* exclude assets folders from coverage report

* added assets

* clean assets

* add source files

* add images files

* add fixtures files

* fix jest failing on svg import

* rename page title

* render hemepage

* cleaned tests

* remove unused import from index page

* Remove extra space

* Add UsersList container, test

fixes #5

* Bg fix style loader (#8)

* add path module

* add build cmd

* add new line

* fix style loader

* Add pagination, user component

* Remove logic for extra links, import connected component

- We don't have 42 pages right now, as we're only getting 100 users
- copy and paste mistake

* Add Grid, styling

* Install semantic ui (#14)

* add path module

* add build cmd

* add new line

* fix style loader

* config files and manual for customizing semantic ui

* keep only changes in site.variables

* removed unused import

* removed unused import from tests

* add semantic folder to gitignore

* add deleted test

* exclude assets folders from coverage report

* added assets

* clean assets

* Remove package-lock

* Readme modification

* Fix styling for users cards

Felistas committed:
Add initial readme instructions (#11)

* Add initial readme instructions

* Add contributing guide instructions

* Minor edits to the contribution.md file

* hot fix

joaopapereira committed:
Remove package-lock.json

Removed because we are using yarn to maintain the dependencies

joaopapereira committed:
Correct README formatting

FedericoEsparza committed:
Merge pull request #29 from AgileVentures/random-changes

Random changes

joaopapereira committed:
Add standardjs (#31)

Correct the projects issues raise by standard
Change the location of test files
Use babel eslint parser with standard

mattwr18 committed:
Add UsersList container, tests

mattwr18 committed:
Add redux to projects, pagination

mattwr18 committed:
Remove default export from ProjectsList

mattwr18 committed:
Connect store, mapStateToProps

mattwr18 committed:
Add redux, pagination

mattwr18 committed:
Gitignore .env

mattwr18 committed:
Add project component (#51)

* Rename css card class for users

joaopapereira commented on src/actions/getProjectsAction.js:
```diff
@@ -0,0 +1,22 @@
+    .then(response => {
+      let { projects, languages, followers, documents } = response.data
+      projects = projects.map(project => {
+        project.languages = languages[project.title] && languages[project.title].map(lang => lang.name)
```
This is starting to get too complicated to be inside this function, mayber we need some helper that translates response to projects list

mattwr18 replied:
you think that is complicated @joaopapereira... this is what I did for the users profile pages https://github.com/AgileVentures/agile-ventures-website-react-front-end/blob/75-add-users-show-page/src/actions/getUsersAction.js 
I'll see if I can extract it to a helper method

joaopapereira commented on src/containers/ProjectsList.js:
```diff
@@ -1,8 +1,208 @@
+      }
+    }
+
+    let languages = new Set()
```
Why do we use a set here?

mattwr18 replied:
so they are a unique list of languages

joaopapereira commented on src/containers/ProjectsList.js:
```diff
@@ -1,8 +1,208 @@
+  }
+
+  normalizeProjects (projects) {
+    let pageCount = Math.ceil(projects.length / 12)
```
This `12` is everywhere in the code
We should get a more elegant way to do this pagination thing

mattwr18 replied:
we could set it as a const somewhere...?

joaopapereira replied:
the const might be a good first step. But due to the amount of places, it feel like something is not 100%

joaopapereira replied:
maybe that is a start. But the fact that is in so many places is telling me that something is not 100%

mattwr18 replied:
we are only using in three places... 1) to calculate the number of total pages based on the number of projects coming in
2) based on the projects paginated by language, 3) based on the number of filtered projects

joaopapereira commented on src/tests/components/Project.test.js:
```diff
@@ -0,0 +1,39 @@
+import projectsFixture from '../../fixtures/projects'
+import fullLogo from '../../images/full_logo2_agile_ventures.png'
+
+// mount Project component
```
do we need this comment?

mattwr18 replied:
probably not... is it not helpful to understand what `mount(<Project />)` does? :stuck_out_tongue:

joaopapereira commented on src/containers/ProjectsList.js:
```diff
@@ -1,8 +1,208 @@
+    }, {})
+  };
+
+  handlePopulateLanguagesDropdown () {
```
should this function be called `populateLanguagesDropdown` ?

mattwr18 replied:
maybe... it seems to be convention to preface these actions with the keyword `handle`, so not sure, can wait to see what people think

Kachulio1 replied:
most people usually start with handle so that people can now the function is for handling an event

Kachulio1 replied:
but for this case, it's not used as an event handler so.

joaopapereira commented on src/containers/ProjectsList.js:
```diff
@@ -1,8 +1,208 @@
+    }
+  }
+
+  normalizeProjects (projects) {
```
This functions is doing too much. The calculation of the languages looks like is not a concern of normalization of projects

mattwr18 replied:
what do you mean by "the calculation of the languages", what we are accomplishing here is going over all the projects and paginating them, then we are going through the languages and adding them as keys with all the projects which use the language as paginated values. 

this allows us to be able to quickly access them when someone filters by language and wants to change pages

joaopapereira replied:
1. Separate the project into pages
2. Check what are the available languages
3. Filter projects
4. Separate them per pages again

These feel like different concerns that 1 single functions should not be doing all. Also the naming I am not 100% sure what `normalize` means

mattwr18 replied:
I agree, this was a first pass at refactoring that @Kachulio1, myself and @dkuku took, but it could definitely be refactored more.

joaopapereira commented on src/containers/ProjectsList.js:
```diff
@@ -1,8 +1,208 @@
+
+  componentDidMount () {
+    if (!this.props.projects.length) {
+      this.props.fetchProjects()
```
if `axios` fails to retrieve the project the error will buble up. Is this what we want?

mattwr18 replied:
I believe what happens is that if the `fetchProjects` is not successful it will just show the spinner... you can go to the /users page right now while the server is off and see this use case

joaopapereira replied:
I didnt try it yet. but I am not sure the intended behavior is for the spinner to be there forever. That tells me that it is trying and even if the server gets back up we will not get the projects.

This PR is already humongus, so maybe we can roll with this and I will create a bug for it

mattwr18 replied:
yeah, sorry... I tried to break it into three parts, but even this wasn't sufficient, and was a result of our building and not putting in `PR`s with small changes last year...

joaopapereira reviewed (changes requested):
Please review my comments

Kachulio1 commented on src/assets/ProjectsList.css:
```diff
@@ -0,0 +1,3 @@
+.search-dropdown {
+  margin: 20px 0px;
+}
```
add a new line

mattwr18 replied:
you add a new line :stuck_out_tongue:

Kachulio1 commented on src/components/Project.js:
```diff
@@ -0,0 +1,33 @@
+  return (
+    <Card className='project-card' raised>
+      <Image
+        src={project.image_url ? project.image_url : fullLogo}
```
You can achieve the same result with less code like this`  project.image_url  | |  fullLogo` 😉

mattwr18 replied:
great! of course you can hehe rookie mistake

Kachulio1 commented:
@mattwr18  can we get a .2 increase

mattwr18 committed:
Standard fixes

mattwr18 committed:
Work through PR review

Co-authored-by: Joseph Ngugi <jngugi88@gmail.com>
- rename normalizeProjects as it's unclear what it was doing
- extract the addLanguagesToProjectsObject function to reduce the complexity and size of its function
- created projectsPerPage constant to clarify what its intended effect is
- rename handlePopulateLanguagesDropdown as its not handling an event
- add newline at end of file
- use more concise either or statement in Project component
- extract updateProjectsObject to reduce complexity of axios request

joaopapereira reviewed (approved):
Going to merge
